File: shared/src/api/services.ts
/** Service operational state */
export type ServiceState = 'not_installed' | 'installed' | 'running' | 'stopped' | 'error';

/** Service information */
export interface ServiceInfo {
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly state: ServiceState;
  readonly version?: string;
  readonly auto_start: boolean;
}

/** AdGuard Home status */
export interface AdGuardStatus {
  readonly enabled: boolean;
  readonly total_queries: number;
  readonly blocked_queries: number;
  readonly block_percentage: number;
  readonly avg_response_ms: number;
}

/** AdGuard DNS forwarding status */
export interface AdGuardDNSStatus {
  readonly enabled: boolean;
  readonly dns_port: number;
}

/** AdGuard Home configuration (raw YAML content) */
export interface AdGuardConfig {
  readonly content: string;
}

/** Type guard for ServiceInfo */
export function isServiceInfo(value: unknown): value is ServiceInfo {
  if (typeof value !== 'object' || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.id === 'string' &&
    typeof v.name === 'string' &&
    typeof v.description === 'string' &&
    typeof v.state === 'string' &&
    typeof v.auto_start === 'boolean'
  );
}

File: backend/internal/models/services.go
package models

// ServiceInfo describes an installable/running service.
type ServiceInfo struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Description string  `json:"description"`
	State       string  `json:"state"`
	Version     *string `json:"version,omitempty"`
	AutoStart   bool    `json:"auto_start"`
}

// AdGuardStatus contains AdGuard Home statistics.
type AdGuardStatus struct {
	Enabled         bool    `json:"enabled"`
	TotalQueries    int64   `json:"total_queries"`
	BlockedQueries  int64   `json:"blocked_queries"`
	BlockPercentage float64 `json:"block_percentage"`
	AvgResponseMS   float64 `json:"avg_response_ms"`
	AdminURL        string  `json:"admin_url,omitempty"`
	ConfigYAMLPath  string  `json:"config_yaml_path,omitempty"`
}

// AdGuardDNSStatus indicates whether AdGuard is configured as the LAN DNS,
// plus health information about the AdGuard DNS listener path.
type AdGuardDNSStatus struct {
	Enabled bool `json:"enabled"`
	DNSPort int  `json:"dns_port"`
	// AdguardListenerReady is true when a TCP connection to AdGuard's DNS port succeeds.
	AdguardListenerReady bool `json:"adguard_listener_ready"`
	// DnsmasqForwardTarget is the current dnsmasq server option (e.g., "127.0.0.1#5353").
	DnsmasqForwardTarget string `json:"dnsmasq_forward_target"`
	// ResolverProbeOk is true when a DNS lookup via the AdGuard listener succeeds.
	ResolverProbeOk bool `json:"resolver_probe_ok"`
}
